app/models/ext_management_system.rb

```diff
@@ -260,6 +264,78 @@ def default_endpoint
     default || endpoints.build(:role => "default")
   end
 
+  # TODO: takes multiple connection data
+  # endpoints, authentications, and works out what's what.
+  def connections=(options)
+    options.each do |option|
+      endpoint = endpoint_builder(option[:endpoint])
+      authentication_builder(option[:authentication])
+        # connection_by_role=(option)
+    end
+  end
+
+  def connections
+    roles = endpoints.map(&:role)
+    options = {}
+
+    roles.each do |role|
+      conn = connection_by_role(role)
+      options[role] = conn
+    end
+
+    connections = OpenStruct.new(options)
+    connections.roles = roles
+    connections
+  end
+
+  # Takes a hash of connection data
+  # hostname, port, and authentication
+  # if no role is passed in assume is default role
+  def connection_by_role=(options)
+    unless options[:endpoint].key?(:role)
+      options[:endpoint][:role] = "default"
+    end
+
+    endpoint_builder(options[:endpoint])
+    authentication_builder(options[:authentication])
+  end
+
+  def endpoint_builder(options)
+    unless options.blank?
+      endpoint = endpoints.detect { |e| e.role == options[:role].to_s }
+      # update or create
+      unless endpoint.nil?
```

Comment:
Prefer flipping the logic here to avoid the confusing "unless/else with a double-negative for good measure" ;)

```ruby
if endpoint
  endpoint.update(options)
else
  endpoints.build(options)
end
```